### PR TITLE
Merge wildcard and CRC32 routines from OpenJ9 into omrutil

### PR DESCRIPTION
**Pull request by mthom**
This commit is in support of the refactored OpenJ9 shared cache into OMR.

Issue: #4610

Signed-off-by: Mark Thom <mark.thom@unb.ca>

### CODE REVIEW

File: util/omrutil/CMakeLists.txt [hunk 1]
```diff
@@ -122,6 +123,7 @@ list(APPEND OBJECTS
 	detectVMDirectory.c
 	gettimebase.c
 	j9memclr.c
+        omrcrc32.c
```
Comment (youngar) [inline util/omrutil/CMakeLists.txt]: change space indentation to tab

File: util/omrutil/wildcard.c [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+		}
+	}
+
+	return retval;
+}
+
+
+
+
+
```
Comment (youngar) [inline util/omrutil/wildcard.c]: Maybe fix some of this whitespace

File: util/omrutil/CMakeLists.txt (hunk 1 above)
Comment (DmonteDD) [reply util/omrutil/CMakeLists.txt]: Thanks for the comment. This change is incorporated in the recent push.

File: include_core/omrutil.h [hunk 3]
```diff
@@ -346,6 +346,53 @@ void OMRZeroMemory(void *ptr, uintptr_t length);
...
+* @brief
+* @param pattern
+* @param patternLength
+* @param needle
+* @param needleLength
+* @param matchFlag
+* @return IDATA
+*/
+IDATA
+parseWildcard(const char * pattern, UDATA patternLength, const char** needle, UDATA* needleLength, U_32 * matchFlag);
```
[the remark this answers is not in the document]
Comment (DmonteDD) [reply include_core/omrutil.h]: Thanks for the comment. I have this incorporated in the recent push.

File: util/omrutil/wildcard.c (hunk 2 above)
Comment (DmonteDD) [reply util/omrutil/wildcard.c]: Thanks for the comment. White spaces are removed now.

File: include_core/omrutil.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (DmonteDD) [reply include_core/omrutil.h]: > I would like to see these prefixed with `omr`. That also means we won't hit a duplicate symbol issue when OpenJ9 links this lib.

Thanks for the comment. I have this incorporated in the recent push.

File: util/omrutil/wildcard.c (hunk 2 above)
Comment (DmonteDD) [reply util/omrutil/wildcard.c]: > Maybe fix some of this whitespace

Thanks for the comment. White spaces are removed now.

File: util/omrutil/CMakeLists.txt (hunk 1 above)
Comment (DmonteDD) [reply util/omrutil/CMakeLists.txt]: > change space indentation to tab

Thanks for the comment. I have this incorporated in the recent push.

**Comment by youngar** [comment]
@genie-omr build all

**Comment by youngar** [comment]
I'm a little concerned about the errors on azure, but I'm not really sure what is wrong.  It looks like the tracegen file is not being processed soon enough.
```
[ 10%] Building C object util/omrutil/CMakeFiles/omrutil_obj.dir/wildcard.c.o
/home/vsts/work/1/s/util/omrutil/wildcard.c:26:24: fatal error: ut_omrutil.h: No such file or directory
compilation terminated.
util/omrutil/CMakeFiles/omrutil_obj.dir/build.make:298: recipe for target 'util/omrutil/CMakeFiles/omrutil_obj.dir/wildcard.c.o' failed
make[2]: *** [util/omrutil/CMakeFiles/omrutil_obj.dir/wildcard.c.o] Error 1

```

The working build has
```
00:00:46  Scanning dependencies of target trc_omrutil
00:00:46  [  7%] Generating ut_omrutil.c, ut_omrutil.h, ut_omrutil.pdat
...

00:00:49  [ 15%] Building C object util/omrutil/CMakeFiles/omrutil_obj.dir/wildcard.c.o
```


The (failing) jenkins linux build looks like its generating the tracefile, but its not happy with the trace functions being used:
```
00:00:25  [131/831] Generating ut_omrutil.c, ut_omrutil.h, ut_omrutil.pdat
...
00:00:25  FAILED: /usr/lib/ccache/cc  -DJ9X86 -DLINUX -D_FILE_OFFSET_BITS=64 -D_X86_ -Iutil/omrutil -I../include_core -I. -I../example/glue -pthread -fno-strict-aliasing -m32 -msse2  -fPIC   -Werror -Wall -MD -MT util/omrutil/CMakeFiles/omrutil_obj.dir/wildcard.c.o -MF util/omrutil/CMakeFiles/omrutil_obj.dir/wildcard.c.o.d -o util/omrutil/CMakeFiles/omrutil_obj.dir/wildcard.c.o   -c /home/jenkins/workspace/Build/util/omrutil/wildcard.c
00:00:25  /home/jenkins/workspace/Build/util/omrutil/wildcard.c: In function ‘omrParseWildcard’:
00:00:25  /home/jenkins/workspace/Build/util/omrutil/wildcard.c:49:2: error: implicit declaration of function ‘Trc_Util_parseWildcard_Entry’ [-Werror=implicit-function-declaration]
00:00:25    Trc_Util_parseWildcard_Entry(patternLength, pattern);
00:00:25    ^
00:00:25  /home/jenkins/workspace/Build/util/omrutil/wildcard.c:66:5: error: implicit declaration of function ‘Trc_Util_parseWildcard_Error’ [-Werror=implicit-function-declaration]
00:00:25       Trc_Util_parseWildcard_Error();
00:00:25       ^
00:00:25  /home/jenkins/workspace/Build/util/omrutil/wildcard.c:75:2: error: implicit declaration of function ‘Trc_Util_parseWildcard_Exit’ [-Werror=implicit-function-declaration]
00:00:25    Trc_Util_parseWildcard_Exit(patternLength, pattern, *matchFlag);
00:00:25    ^
00:00:25  cc1: all warnings being treated as errors
00:00:25  ninja: build stopped: subcommand failed.
```

**Comment by DmonteDD** [comment]
> I'm a little concerned about the errors on azure, but I'm not really sure what is wrong. It looks like the tracegen file is not being processed soon enough.
> 
> ```
> [ 10%] Building C object util/omrutil/CMakeFiles/omrutil_obj.dir/wildcard.c.o
> /home/vsts/work/1/s/util/omrutil/wildcard.c:26:24: fatal error: ut_omrutil.h: No such file or directory
> compilation terminated.
> util/omrutil/CMakeFiles/omrutil_obj.dir/build.make:298: recipe for target 'util/omrutil/CMakeFiles/omrutil_obj.dir/wildcard.c.o' failed
> make[2]: *** [util/omrutil/CMakeFiles/omrutil_obj.dir/wildcard.c.o] Error 1
> ```
> 
> The working build has
> 
> ```
> 00:00:46  Scanning dependencies of target trc_omrutil
> 00:00:46  [  7%] Generating ut_omrutil.c, ut_omrutil.h, ut_omrutil.pdat
> ...
> 
> 00:00:49  [ 15%] Building C object util/omrutil/CMakeFiles/omrutil_obj.dir/wildcard.c.o
> ```
> 
> The (failing) jenkins linux build looks like its generating the tracefile, but its not happy with the trace functions being used:
> 
> ```
> 00:00:25  [131/831] Generating ut_omrutil.c, ut_omrutil.h, ut_omrutil.pdat
> ...
> 00:00:25  FAILED: /usr/lib/ccache/cc  -DJ9X86 -DLINUX -D_FILE_OFFSET_BITS=64 -D_X86_ -Iutil/omrutil -I../include_core -I. -I../example/glue -pthread -fno-strict-aliasing -m32 -msse2  -fPIC   -Werror -Wall -MD -MT util/omrutil/CMakeFiles/omrutil_obj.dir/wildcard.c.o -MF util/omrutil/CMakeFiles/omrutil_obj.dir/wildcard.c.o.d -o util/omrutil/CMakeFiles/omrutil_obj.dir/wildcard.c.o   -c /home/jenkins/workspace/Build/util/omrutil/wildcard.c
> 00:00:25  /home/jenkins/workspace/Build/util/omrutil/wildcard.c: In function ‘omrParseWildcard’:
> 00:00:25  /home/jenkins/workspace/Build/util/omrutil/wildcard.c:49:2: error: implicit declaration of function ‘Trc_Util_parseWildcard_Entry’ [-Werror=implicit-function-declaration]
> 00:00:25    Trc_Util_parseWildcard_Entry(patternLength, pattern);
> 00:00:25    ^
> 00:00:25  /home/jenkins/workspace/Build/util/omrutil/wildcard.c:66:5: error: implicit declaration of function ‘Trc_Util_parseWildcard_Error’ [-Werror=implicit-function-declaration]
> 00:00:25       Trc_Util_parseWildcard_Error();
> 00:00:25       ^
> 00:00:25  /home/jenkins/workspace/Build/util/omrutil/wildcard.c:75:2: error: implicit declaration of function ‘Trc_Util_parseWildcard_Exit’ [-Werror=implicit-function-declaration]
> 00:00:25    Trc_Util_parseWildcard_Exit(patternLength, pattern, *matchFlag);
> 00:00:25    ^
> 00:00:25  cc1: all warnings being treated as errors
> 00:00:25  ninja: build stopped: subcommand failed.
> ```

Yes, I saw checked these errors. I guess, I need to add **ut_omrutil.c** in **omr_add_library** in CMakeLists.txt. I will change that and check if thats the cause.

**Comment by DmonteDD** [comment]
@youngar can you please _build all_. Thanks

**Comment by youngar** [comment]
@genie-omr build all

**Comment by youngar** [comment]
Looks like adding the source to the file to the library will properly set up the dependency to generate the source and header file. https://github.com/eclipse/omr/blob/master/cmake/modules/OmrTracegen.cmake#L65
Maybe in the future we should make the `add_tracegen` function more featureful.

**Pull request merged**